### PR TITLE
fix(plot): Fix dataset_id invalidation in encoding and metadata methods

### PR DESCRIPTION
**Pull request by lmeyerov**
## Summary
Fixed dataset_id invalidation logic for encoding, metadata, and style methods. Dataset IDs now properly invalidate when encodings, name, description, or visualization styles change, preventing stale ID reuse after modifying visualization properties.

## Analysis Approach
Used complementary analysis tools to ensure complete coverage:
- **AST pattern matching**: Found 9 methods with direct attribute modifications
- **Pysa call graph analysis**: Found 11 methods total, including 2 additional bugs AST missed (+17% bug detection)

## Changes
**Fixed 7 bugs** in `graphistry/PlotterBase.py`:
1. `__encode()` helper - Now invalidates dataset_id (fixes all 8 encoding methods)
2. `encode_axis()` - Now invalidates dataset_id
3. `name()` - Now invalidates dataset_id  
4. `description()` - Now invalidates dataset_id
5. `bind()` - Conditional invalidation (only when changing encodings, not when setting IDs)
6. **`style()` - Now invalidates dataset_id (Pysa-discovered)**
7. **`addStyle()` - Now invalidates dataset_id (Pysa-discovered)**

**Why Pysa Found More:**
- AST finds direct patterns: `res._attr = value`
- Pysa finds indirect patterns: `res = self.bind()` followed by modifications
- `style()` and `addStyle()` use double-copy pattern via `bind()` that AST missed

## Test Coverage
**Comprehensive test suite** with 25 tests in `graphistry/tests/test_dataset_id_invalidation.py`:
- ✅ 11 core tests for basic invalidation scenarios
- ✅ 3 Pysa-discovered bug tests (style, addStyle, infer_labels)
- ✅ 8 parametric tests covering ALL encoding methods (future-proof)
- ✅ 3 bind() edge case tests (multiple encodings, no params, mixed usage)

**Test design principles:**
- Parametric over individual (one test covers all 8 encoding methods)
- Future-proof (adding new encoding = add to parametrize list)
- Non-brittle (clear assertions, independent tests, representative samples)

## Validation
- ✅ All 25 tests passing (Python 3.8 and 3.12)
- ✅ PlotterBase tests passing (48 passed, no regressions)
- ✅ Type checking clean (mypy)
- ✅ Linting clean (flake8)

## Impact
Server-side uploads now correctly detect when datasets have changed, preventing visualization artifacts from stale data. Multi-tool analysis (AST + Pysa) provided 17% better bug detection than AST alone.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ai/prompts/PYRE_ANALYSIS.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
+# Pyre Analysis for PyGraphistry
+
+Pyre-check is a performant type checker for Python that can also be used for advanced code analysis and refactoring. This guide shows how to use pyre on PyGraphistry for tasks beyond what grep/LLM can handle.
+
+## When to Use Pyre vs Grep vs AST Scripts
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: shorten intro + this

File: ai/prompts/PYRE_ANALYSIS.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+- Custom pattern detection (e.g., "methods that modify attribute X")
+- Pyre is too slow or times out
+- Need fast iteration during development
+
+**Use Grep When:**
+- Simple string/pattern search
+- Quick file location
+- Initial exploration
+
+## Installation
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: this can be shorter

File: ai/prompts/PYRE_ANALYSIS.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+uv tool install --python python3.12 pyre-check
+
+# Verify installation
+which pyre  # Should show: /home/lmeyerov/.local/bin/pyre
+pyre --version
+```
+
+## Configuration
+
+PyGraphistry has a `.pyre_configuration` file:
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: just give config file's path

File: ai/prompts/PYRE_ANALYSIS.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+```
+
+## Running Pyre
+
+### Option 1: Docker (Recommended - Avoids Library Issues)
+
+```bash
+# Using python:3.12-slim image (has required glibc/libstdc++ versions)
+docker run --rm -it \
+    -v /home/lmeyerov/Work/pygraphistry2:/workspace:ro \
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: use `pwd` not wrong abs path

File: ai/prompts/PYRE_ANALYSIS.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+**Lesson**: For custom analysis patterns, AST scripts are often more practical than pyre queries.
+
+## Troubleshooting
+
+### Pyre Times Out
+
+**Symptom**: Hangs at "Processed 1631 of 1709 functions" for 5+ minutes.
+
+**Solutions**:
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md:51]: if timeout, turn on verbose/debug mode to idnetify problematic method and see if can solve such as fixing file or ignoring it

File: ai/prompts/PYRE_ANALYSIS.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+| Type checking | ✅ Yes | ✅ Yes |
+| Speed | Slower (5-10min) | Faster (< 1min) |
+| Query API | ✅ Yes | ❌ No |
+| Call graphs | ✅ Yes | ❌ No |
+| Find references | ✅ Yes | ❌ No |
+| PyGraphistry default | ❌ No | ✅ Yes |
+
+**Recommendation**: Use mypy for regular type checking, pyre for advanced refactoring analysis.
+
+## Future Work
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: remove future work

File: ai/prompts/PYRE_ANALYSIS.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+- **Performance**: Investigate pyre incremental mode for faster analysis
+- **Integration**: Add pyre query scripts for common refactoring tasks
+- **Documentation**: Document useful query patterns as they're discovered
+- **CI/CD**: Consider pyre for static analysis in CI (if performance improves)
+
+## Resources
+
+- [Pyre Documentation](https://pyre-check.org/)
+- [Pyre Query API](https://pyre-check.org/docs/querying-pyre/)
+- PyGraphistry AST Scripts: `plans/*/analyze_*.py`
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: ./plans/ is .gitignored, move analyze_*.py reference script into ai/ folder somewhere (assets?)

File: ai/prompts/PYRE_ANALYSIS.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
+# Pyre/Pysa Analysis for PyGraphistry
+
+**Status**: ✅ Pysa found 2 bugs AST missed (2025-10-16, Pyre v0.9.25)
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: why is there a status here? this is supposed to be a workflow for others

File: ai/prompts/PYRE_ANALYSIS.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+**Extract callers** (see `ai/assets/pysa_extract_callers.py`):
+```bash
+python3 ai/assets/pysa_extract_callers.py pysa_results/call-graph.json \
+  PlotterBase.PlotterBase.bind \
+  PlotterBase.PlotterBase.nodes
+```
+
+**Config**: `.pyre_configuration` (excludes hop.py to prevent hang)
+
+## Pysa Workflow
```
Comment (lmeyerov) [inline ai/prompts/PYRE_ANALYSIS.md]: i thought we had a section on top commands?

File: graphistry/compute/hop.py [hunk 10]
```diff
@@ -1,3 +1,12 @@
+"""
+Graph hop/traversal operations for PyGraphistry
+
+NOTE: This file is excluded from pyre type checking (.pyre_configuration)
+Reason: The hop() function (384 lines, complex control flow) exceeds pyre's
+        analysis complexity limit, causing pyre to hang indefinitely.
+        Use mypy for type checking this file instead.
+See: plans/fix_dataset_file_id_invalidation/pyre_type_hints_proposal.md
+"""
```
Comment (lmeyerov) [inline graphistry/compute/hop.py:5]: Pyre NOTE could have been 1-2 short lines, not 5 long ones
